src/exporter/DaemonMetricCollector.cc

```diff
@@ -0,0 +1,202 @@
+#include "DaemonMetricCollector.h"
+#include "common/admin_socket_client.h"
+#include "common/perf_counters.h"
+#include "include/common_fwd.h"
+
+#include <boost/json/src.hpp>
+#include <filesystem>
+#include <iostream>
+#include <regex>
+#include <string>
+#include <utility>
+
+using json_object = boost::json::object;
+using json_value = boost::json::value;
+using json_array = boost::json::array;
+
+const char *DaemonMetricCollector::SOCKETDIR = "/var/run/ceph/";
+
+void DaemonMetricCollector::request_loop(boost::asio::deadline_timer &timer) {
+	timer.async_wait([&](const boost::system::error_code &e) {
+		std::cerr << e << std::endl;
+		update_sockets();
+		dump_asok_metrics();
+		timer.expires_from_now(boost::posix_time::seconds(stats_period));
+		request_loop(timer);
+	});
+}
+
+void DaemonMetricCollector::main() {
+	// TODO: let's do 5 for now and expose this to change in the future
+	stats_period = 5;
+	boost::asio::io_service io;
+	boost::asio::deadline_timer timer(io,
+																		boost::posix_time::seconds(stats_period));
+	request_loop(timer);
+	io.run();
+}
+
+std::string DaemonMetricCollector::get_metrics() { return metrics; }
+
+template <class T>
+void add_metric(std::stringstream &ss, T value, std::string name,
+								std::string description, std::string mtype,
+								std::string labels) {
+	ss << "# HELP " << name << " " << description << "\n";
+	ss << "# TYPE " << name << " " << mtype << "\n";
+	ss << name << "{" << labels << "} " << value << "\n";
+}
+
+void add_double_or_int_metric(std::stringstream &ss, json_value value,
+															std::string name, std::string description,
+															std::string mtype, std::string labels) {
+	if (value.is_int64()) {
```

Comment:
Tabs look weird here tbh :/

Reply:
Formatted the files using [C/C++ ](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) for VS Code